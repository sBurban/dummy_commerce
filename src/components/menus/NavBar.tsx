import {useState} from 'react'
import { Box,Typography,Toolbar,Link,IconButton,Badge,Paper,InputBase } from '@mui/material';
import AppBar from '@mui/material/AppBar';

import {
    // Menu as MenuIcon, Mail as MailIcon, Notifications as NotifyIcon,
    AccountCircle, Search as SearchIcon, MoreVert as MoreIcon,
    ShoppingCart as ShoppingCartIcon
} from '@mui/icons-material';

import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';


import { useSession, signIn, signOut } from "next-auth/react"
import { useRouter } from 'next/router';
import Image from 'next/image';
import { styled, alpha } from '@mui/material/styles';

const SearchParent = styled('div')(({ theme }) => ({
  position: 'relative',
  borderRadius: theme.shape.borderRadius,
  display: "flex",
  backgroundColor: alpha(theme.palette.common.white, 0.15),
  '&:hover': {
    backgroundColor: alpha(theme.palette.common.white, 0.25),
  },
  marginRight: theme.spacing(2),
  marginLeft: 0,
  width: '100%',
  maxWidth: '600px',
  [theme.breakpoints.up('sm')]: {
    marginLeft: theme.spacing(3),
    // width: 'auto',
  },
}));

const SearchIconWrapper = styled('div')(({ theme }) => ({
  padding: theme.spacing(0, 2),
  height: '100%',
  position: 'absolute',
  pointerEvents: 'none',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
}));

const StyledInputBase = styled(InputBase)(({ theme }) => ({
  color: 'inherit',
  width: '100%',
  '& .MuiInputBase-input': {
    padding: theme.spacing(1, 1, 1, 0),
    // vertical padding + font size from searchIcon
    paddingLeft: `calc(1em + ${theme.spacing(4)})`,
    transition: theme.transitions.create('width'),
    width: '100%',
    [theme.breakpoints.up('md')]: {
      width: '20ch',
    },
  },
}));

const Img = styled('img')({
    // margin: 'auto',
    // display: 'block',
    // maxWidth: '100%',
    // maxHeight: '100%',
    // borderRadius: "50%",
    // backgroundColor: "#fff",
});

export const NavBar = ({onSearchSubmit, ...props}:any) => {
  const [searchText, setSearchText] = useState("");
  const handleText = (e:React.ChangeEvent<HTMLInputElement>) => {
    const target = e.currentTarget;
    // console.log("ðŸš€ ~ file: NavBar.tsx:73 ~ handleText ~ target:", target.value)
    setSearchText(target.value);
  }

// console.log("ðŸš€ ~ file: NavBar.tsx:15 ~ NavBar ~ props:", props)
    const router = useRouter();
    const { data: session, status } = useSession();
    // console.log("ðŸš€ ~ file: NavBar.tsx:31 ~ NavBar ~ session:", session)

    const [anchorElemem, setAnchorElemem] = useState<any|null>(null);
    const [mobileMoreAnchorEl, setMobileMoreAnchorEl] = useState<any|null>(null);
    const isMenuOpen = Boolean(anchorElemem);
    const isMobileMenuOpen = Boolean(mobileMoreAnchorEl);
    const handleMenuClose = () => {
        setAnchorElemem(null);
    };
    const handleProfileMenuOpen = (event:any) => {
        setAnchorElemem(event.currentTarget);
    };
    const handleMobileMenuOpen = (event:any) => {
        setMobileMoreAnchorEl(event.currentTarget);
    };
    const handleMobileMenuClose = () => {
        setMobileMoreAnchorEl(null);
    };


    const handleAccountButton = () => {
        router.push('/account');
        handleMenuClose();
    }
    const handleProfileButton = () => {
        router.push('/account/profile');
        handleMenuClose();
    }


    const profileElement = !session? <AccountCircle />
    : <Paper sx={{ width: '3rem', height: '3rem' }}>
      <Img alt="profile image" src={session?.user?.image || "#"} />
    </Paper>


    const menuId = 'navbar-account-menu';
    const renderProfileMenu = (
        <Menu
          anchorEl={anchorElemem}
          anchorOrigin={{
            vertical: 'top',
            horizontal: 'right',
          }}
          id={menuId}
          keepMounted
          transformOrigin={{
            vertical: 'top',
            horizontal: 'right',
          }}
          open={isMenuOpen}
          onClose={handleMenuClose}
        >
          <MenuItem onClick={handleAccountButton}>Account</MenuItem>
          <MenuItem onClick={handleProfileButton}>Profile</MenuItem>
          {session?
            <MenuItem onClick={() => signOut()}>Sign Out</MenuItem>
            :<MenuItem onClick={() => signIn()}>Sign In</MenuItem>
          }

        </Menu>
    );



    const mobileMenuId = 'navbar-account-menu-mobile';
    const renderMobileProfileMenu = (
        <Menu
          anchorEl={mobileMoreAnchorEl}
          anchorOrigin={{
            vertical: 'top',
            horizontal: 'right',
          }}
          id={mobileMenuId}
          keepMounted
          transformOrigin={{
            vertical: 'top',
            horizontal: 'right',
          }}
          open={isMobileMenuOpen}
          onClose={handleMobileMenuClose}
        >

          <MenuItem>
            <IconButton
              size="large"
              aria-label="show 4 items in cart"
              color="inherit"
            >
              <Badge badgeContent={4} color="error">
                <ShoppingCartIcon />
              </Badge>
            </IconButton>
            <p>Shopping Cart</p>
          </MenuItem>

          <MenuItem onClick={handleProfileMenuOpen}>
            <IconButton
              size="large"
              aria-label="account of current user"
              aria-controls="primary-search-account-menu"
              aria-haspopup="true"
              color="inherit"
            >
              {profileElement}
            </IconButton>
            <p>Profile</p>
          </MenuItem>


        </Menu>
      );




    return (
        <Box sx={{ flexGrow: 1 }} className='Navbar' >
          <AppBar position="static">
            <Toolbar>

                {/* Left side of the NavBar */}

                <Link href={"/"}>
                    <Typography
                        variant="h6"
                        noWrap
                        component="div"
                        sx={{ display: { xs: 'none', sm: 'block', color: '#fff' } }}
                    >
                        Duhmmerce
                    </Typography>
                </Link>

                <SearchParent className='search_box_parent'>
                    {/* <SearchIconWrapper>
                        <SearchIcon />
                    </SearchIconWrapper> */}
                    <StyledInputBase
                        placeholder="Searchâ€¦"
                        inputProps={{ 'aria-label': 'search' }}
                        value={searchText}
                        onChange={handleText}
                    />
                    <IconButton
                        // size="large"
                        aria-label="searchbar button"
                        sx={{
                          backgroundColor: "#febd69",
                          color: "#333",
                          borderRadius: "unset",
                          '&:hover':{
                            backgroundColor: "#f3a847"
                          }
                        }}
                        onClick={() => onSearchSubmit(searchText)}
                    >
                      <SearchIcon />
                    </IconButton>
                </SearchParent>


                {/* Creates space between left and right of the NavBar */}
                <Box sx={{ flexGrow: 1 }} />


                {/* Right-side of NavBar. Starts here */}

                {/* Box displayed on bigger screen sizes. Controls normal menu*/}
                <Box sx={{ display: { xs: 'none', md: 'flex' } }}>
                    <IconButton
                        size="large"
                        aria-label="show 4 items in cart"
                        color="inherit"
                    >
                        <Badge badgeContent={4} color="error">
                        <ShoppingCartIcon />
                        </Badge>
                    </IconButton>

                    <IconButton
                        size="large"
                        edge="end"
                        aria-label="account of current user"
                        aria-controls={menuId}
                        aria-haspopup="true"
                        onClick={handleProfileMenuOpen}
                        color="inherit"
                    >
                        {profileElement}
                    </IconButton>
                </Box>

                {/* Box displayed on smaller screen sizes. Controls Mobile Menu */}
                <Box sx={{ display: { xs: 'flex', md: 'none' } }}>
                    <IconButton
                    size="large"
                    aria-label="show more"
                    aria-controls={mobileMenuId}
                    aria-haspopup="true"
                    onClick={handleMobileMenuOpen}
                    color="inherit"
                    >
                    <MoreIcon />
                    </IconButton>
                </Box>

            </Toolbar>
          </AppBar>
          {renderMobileProfileMenu}
          {renderProfileMenu}
        </Box>
    );
}
