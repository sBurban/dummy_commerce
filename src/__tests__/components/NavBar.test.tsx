import React from "react";
import {render, screen, } from '@testing-library/react';
import NavBar from "../../components/menus/NavBar";

//https://github.com/vercel/next.js/discussions/23034 - ghost on Dec 23, 2019
jest.mock("next/router", () => ({
    useRouter() {
      return {
        route: "",
        pathname: "",
        query: "",
        asPath: "",
      };
    },
  }));
const useRouter = jest.spyOn(require('next/router'), 'useRouter')

//https://github.com/nextauthjs/next-auth/discussions/4185 - agangdi on Mar 20, 2022

jest.mock("next-auth/react", () => {
    const originalModule = jest.requireActual('next-auth/react');
    const mockSession = {
      expires: new Date(Date.now() + 2 * 86400).toISOString(),
      user: { username: "admin" }
    };
    return {
      __esModule: true,
      ...originalModule,
      useSession: jest.fn(() => {
        return {data: mockSession, status: 'authenticated'}  // return type is [] in v3 but changed to {} in v4
      }),
    };
});

describe('NavBar component', () => {
    test('renders without errors', () => {
        useRouter.mockImplementation(() => ({
            route: "/yourRoute",
            pathname: "/yourRoute",
            query: "",
            asPath: "",
        }));

        const { container } = render(<NavBar />);
        expect(container).toBeInTheDocument();
        expect(container).toMatchSnapshot();

    });

    test('render static Texts', async () => {
        const { container } = render(<NavBar />);
        // expect(container).toBeInTheDocument();
        const textElements = await screen.findAllByText("About us", {exact: false});
        const expectedElems = textElements.length;
        expect(textElements).toHaveLength(expectedElems);
    })

    // test('show "sign out" when has session', async () => {
    //     const {container} = render(<NavBar/>);
    //     expect(container).toMatchSnapshot()
    //     expect(screen.getByText("Sign Out")).toBeInTheDocument();
    // })

});


