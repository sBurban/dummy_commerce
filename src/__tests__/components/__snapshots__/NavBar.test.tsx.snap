// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar component renders without errors 1`] = `
<div>
  <div
    class="Navbar MuiBox-root css-i9gxme"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      >
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1bhi0za-MuiTypography-root-MuiLink-root"
          href="/"
        >
          <div
            class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-b9ntii-MuiTypography-root"
          >
            Duhmmerce
          </div>
        </a>
        <div
          class="search_box_parent css-136to22"
        >
          <div
            class="MuiInputBase-root MuiInputBase-colorPrimary css-zme7u2-MuiInputBase-root"
          >
            <input
              aria-label="search"
              class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
              placeholder="Searchâ€¦"
              type="text"
              value=""
            />
          </div>
          <button
            aria-label="searchbar button"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-z5mkrl-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-i9gxme"
        />
        <div
          class="MuiBox-root css-1guk29"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1bhi0za-MuiTypography-root-MuiLink-root"
              href="/about_us/"
            >
              <div
                class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-b9ntii-MuiTypography-root"
              >
                About us
              </div>
            </a>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
          />
          <button
            aria-label="show 4 items in cart"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiBadge-root css-1c32n2y-MuiBadge-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ShoppingCartIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorError css-to4x1n-MuiBadge-badge"
              >
                4
              </span>
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-controls="navbar-account-menu"
            aria-haspopup="true"
            aria-label="account of current user"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeEnd MuiIconButton-sizeLarge css-1f5y6oa-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1th43wq-MuiPaper-root"
            >
              <img
                alt="profile image"
                class="css-y9s1zk"
                src="#"
              />
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-1m04nb5"
        >
          <button
            aria-controls="navbar-account-menu-mobile"
            aria-haspopup="true"
            aria-label="show more"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MoreVertIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </header>
  </div>
</div>
`;
